plugins/formal-verification/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt

```diff
@@ -70,36 +70,46 @@ class StmtConversionVisitor : FirVisitor<Exp, StmtConversionContext>() {
             else -> TODO("Constant Expression of type ${constExpression.kind} is not yet implemented.")
         }
 
-    override fun visitWhenExpression(whenExpression: FirWhenExpression, data: StmtConversionContext): Exp {
-        if (whenExpression.branches.size != 2) {
-            // For now only when expressions with 2 branches are supported so if statements can be encoded.
-            // When expressions with more than 2 branches can be embedded as nested if-else chains.
-            TODO("When expressions with more that two branches are not yet implemented")
-        }
+    override fun visitWhenSubjectExpression(whenSubjectExpression: FirWhenSubjectExpression, data: StmtConversionContext): Exp {
+        return whenSubjectExpression.whenRef.value.subject?.accept(this, data)
```

Comment:
Unfortunately I cannot find any example in which this happens. But since the subject is nullable I have to handle the case

Reply:
Ah, I see what's going on.  The `FirWhenSubjectExpression` node is for when you use the subject (implicitly), but to actually get the subject you need to go through the when expression again... a bit inconvenient, that.
Could you check whether this works correctly for subjects with a side effect?

Reply:
I talked with Ilya about subjects implicit calls and we can just do an unsafe cast here. Anyway if the subject is a function call (which may have side effects) there's already an example that i I think is wrong.
```
fun <!VIPER_TEXT!>when_with_subject_call<!>(x: Int): Int {
    return when (when_with_subject_var(x)) {
        1 -> 2
        2 -> 3
        else -> 42
    }
}
```
which is translated in this way
```
method global$pkg_$when_with_subject_call(local$x: Int) returns (ret$: Int)
{
  var anonymous$1: Int
  var anonymous$2: Int
  anonymous$2 := global$pkg_$when_with_subject_var(local$x)
  if (anonymous$2 == 1) {
    anonymous$1 := 2
  } else {
    var anonymous$3: Int
    anonymous$3 := global$pkg_$when_with_subject_var(local$x)
    if (anonymous$3 == 2) {
      anonymous$1 := 3
    } else {
      anonymous$1 := 42}
  }
  ret$ := anonymous$1
}
```
Fixing this will not be so easy :)

Reply:
I think an unsafe cast makes sense.  The issue with the subject being evaluated multiple times is a problem, though.  I'll think about how to resolve that.  Can you make sure to include a TODO on it?